core/src/main/scala/kafka/log/LogCleaner.scala

```diff
@@ -612,12 +617,23 @@ private[log] class Cleaner(val id: Int,
       if (outputBuffer.position() > 0) {
         outputBuffer.flip()
         val retained = MemoryRecords.readableRecords(outputBuffer)
+        val baseOffsetOfLog = dest.baseOffset
+        val largestOffsetToAppend = result.maxOffset
+
+        if ((largestOffsetToAppend - baseOffsetOfLog) > Integer.MAX_VALUE) {
+          require(numSegmentsInGroup == 1, s"Constructed segment group causes index offset overflow for $topicPartition")
```

Comment:
Can we add a comment about this check since it's not obvious at a glance why the number of grouped segments has anything to do with overflow?